### PR TITLE
Read `SourceConfig` from file

### PR DESCRIPTION
**Pull request by guilload**
### Description
Read `SourceConfig` from file. 

### How was this PR tested?
- added a unit test
- indexed data with Kafka source loading the Kafka source conf from local file

### CODE REVIEW

File: quickwit-indexing/src/source/mod.rs [hunk 1]
```diff
@@ -161,7 +162,7 @@ pub fn quickwit_supported_sources() -> &'static SourceLoader {
     })
 }
 
-#[derive(Clone)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
```
Comment (fulmicoton) [inline quickwit-indexing/src/source/mod.rs:165]: PartialEq might be a little ambitious... You'd be surprised by the semantics of PartialEq of a json object in general, and `serde_json::Value` in particular.

File: quickwit-cli/src/lib.rs [hunk 2]
```diff
@@ -220,15 +223,40 @@ pub async fn index_data_cli(args: IndexDataArgs) -> anyhow::Result<()> {
...
-    SourceConfig {
-        id: "cli_source".to_string(),
-        source_type: "file".to_string(),
-        params: serde_json::to_value(params).unwrap(),
+async fn create_source_config_from_args(
+    source_config_path_opt: Option<&PathBuf>,
+    input_path_opt: Option<&PathBuf>,
+) -> anyhow::Result<SourceConfig> {
+    if let Some(source_config_path) = source_config_path_opt {
+        let source_config_file = File::open(source_config_path).with_context(|| {
```
Comment (fulmicoton) [inline quickwit-cli/src/lib.rs:231]: nitpick.
`fs::read` works fine for small files. It avoid building the `BufReader`.

Another nice thing is that your entirely separate the io from the deserialization, making the source of the error clearer.
https://doc.rust-lang.org/stable/std/fs/fn.read.html

File: quickwit-cli/src/lib.rs [hunk 3]
```diff
@@ -220,15 +223,40 @@ pub async fn index_data_cli(args: IndexDataArgs) -> anyhow::Result<()> {
...
-    };
-    SourceConfig {
-        id: "cli_source".to_string(),
-        source_type: "file".to_string(),
-        params: serde_json::to_value(params).unwrap(),
+async fn create_source_config_from_args(
+    source_config_path_opt: Option<&PathBuf>,
+    input_path_opt: Option<&PathBuf>,
+) -> anyhow::Result<SourceConfig> {
+    if let Some(source_config_path) = source_config_path_opt {
```
Comment (fulmicoton) [inline quickwit-cli/src/lib.rs:230]: We should return an error if both Option are "Some"

File: quickwit-cli/src/lib.rs [hunk 4]
```diff
@@ -220,15 +223,40 @@ pub async fn index_data_cli(args: IndexDataArgs) -> anyhow::Result<()> {
...
 
-fn create_source_config_from_args(input_path_opt: Option<PathBuf>) -> SourceConfig {
-    let params = FileSourceParams {
-        filepath: input_path_opt,
-    };
-    SourceConfig {
-        id: "cli_source".to_string(),
-        source_type: "file".to_string(),
-        params: serde_json::to_value(params).unwrap(),
+async fn create_source_config_from_args(
```
Comment (fulmicoton) [inline quickwit-cli/src/lib.rs:226]: One line or too of doc to explain the intent could be nice.

File: quickwit-cli/src/lib.rs [hunk 5]
```diff
@@ -492,6 +520,60 @@ impl ThroughputCalculator {
...
 
+#[cfg(test)]
+mod tests {
+    use serde_json::json;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_create_source_config_from_args() -> anyhow::Result<()> {
+        {
```
Comment (fulmicoton) [inline quickwit-cli/src/lib.rs:531]: nitpick. 2 unit tests would have been nicer.

**Review by fulmicoton — CHANGES_REQUESTED** [review]
Can you put some doc somewhere with an example conf?

Probably the actually documentation is the best place.
(It is not published automatically so don't worry)

Another place is ok too... The main point is to save other devs 10mn of reverse engineering.

**Comment by guilload** [comment]
I made the changes you requested. For now, I added the doc to `SourceConfig`. I'll update our primary documentation with all the updates right before our next release.

Finally, I took the liberty to rename `SourceConfig.id` to `SourceConfig.source_id` to be more consistent with our naming conventions (`index_id`, `split_id`).

File: quickwit-indexing/src/source/mod.rs [hunk 6]
```diff
@@ -161,9 +162,27 @@ pub fn quickwit_supported_sources() -> &'static SourceLoader {
     })
 }
 
-#[derive(Clone)]
+/// A `SourceConfig` describes the properties of a source. A source config can be created
+/// dynamically or loaded from a file consisting of a JSON object with 3 mandatory properties:
+/// - `source_id`, a name identifying the source uniquely;
+/// - `source_type`, the type of the target source, for instance, `kafka` or `kinesis`;
```
Comment (fmassot) [inline quickwit-indexing/src/source/mod.rs:168]: do we have a working kinesis source?

File: quickwit-indexing/src/source/mod.rs [hunk 7]
```diff
@@ -161,9 +162,27 @@ pub fn quickwit_supported_sources() -> &'static SourceLoader {
...
+///         "group_id": "my-kafka-source-consumer-group",
+///         "topic": "my-kafka-source-topic"
+///     }
+/// }
+/// ```
+#[derive(Clone, Serialize, Deserialize)]
 pub struct SourceConfig {
-    pub id: String,
+    pub source_id: String,
     pub source_type: String,
```
Comment (fmassot) [inline quickwit-indexing/src/source/mod.rs:186]: it seems strange to have a source type as a `String`.

**Review by fmassot — APPROVED** [review]

File: quickwit-indexing/src/source/mod.rs (hunk 6 above)
Comment (guilload) [reply quickwit-indexing/src/source/mod.rs:168]: No, not yet. I'll remove the mention to Kinesis.

**Pull request merged**